File: Javascript/Algorithms/500.js
// Only make sets of upper case chars.
let make_rows = () => {
    const rows = ["QWERTYUIOP", "ASDFGHJKL", "ZXCVBNM"];
    let s = new Set;
    for (let r of rows){
        s.add(new Set(r));
    }
    return s;
}

// Returns true if all elements of s1 belong
// in s2.
let issubset = (s1, s2) => {
    for(el of s1){
        if (!s2.has(el)){
            return false;
        }
    }
    return true;
}

var findWords = function(words) {
    const result = [],
          rows = make_rows();
    for (let w of words){
        let sw = new Set(w.toUpperCase());
        for (r of rows){
            if (issubset(sw, r)){
                result.push(w);
            }
        }
    }
    return result;
};

File: Rust/Algorithms/500.rs
use std::collections::HashSet;
use std::iter::FromIterator;

impl Solution {
    fn make_rows() -> Vec<HashSet<char>> {
        let mut rows = Vec::with_capacity(3);
        for &row in &["QWERTYUIOP", "ASDFGHJKL", "ZXCVBNM"] {
            rows.push(HashSet::from_iter(row.chars()));
        }
        return rows;
    }

    pub fn find_words(mut words: Vec<String>) -> Vec<String> {
        let rows = Solution::make_rows();
        let mut result = Vec::new();

        for word in words.iter_mut() {
            let wset = HashSet::from_iter(word.to_uppercase().chars());
            for set in rows.iter() {
                if wset.is_subset(&set) {
                    result.push(word.to_owned());
                }
            }
        }
        result
    }
}
